File: mystock/web/static/data-status.js
/* Read-only provenance display; times stay in their source/UTC zone. */
const cacheLabels = {ok:'采集成功', partial:'部分数据缺失', empty:'本次未返回数据', error:'采集失败', unsupported:'来源暂不支持', unknown:'无法判断', current:'已覆盖最近收盘日', stale:'缓存滞后', pending:'当日尚未确认收盘', awaiting_final_data:'等待收盘后采集', cached:'已有缓存'};
function cacheLabel(status) { return cacheLabels[status] || '无法判断'; }
function cacheTime(value) {
  if (!value) return '未知';
  const known = /(?:Z|[+-]\d{2}:\d{2})$/.test(value);
  return esc(value.replace('T',' ').replace(/(?:Z|\+00:00)$/, ' UTC')) + (known ? '' : '（时区未知）');
}
function cacheNotice(status) { return ['ok','current','cached','pending'].includes(status) ? '' : ' cache-warning'; }
function cacheDaily(d) {
  return `<div class="cache-state${cacheNotice(d.status)}"><b>日线 · ${esc(cacheLabel(d.status))}</b>
    <p>行情日期 ${esc(d.data_as_of || '未知')} · 应覆盖收盘日 ${esc(d.expected_session || '无法判断')}</p>
    <p>该股采集 ${cacheTime(d.collected_at)} · 最近来源尝试 ${esc(cacheLabel(d.attempt_status))}（${d.scope === 'stock' ? '该股' : '来源汇总，不代表逐股成功'}）</p><details><summary>yfinance 日线采集说明</summary><p>最近尝试 ${cacheTime(d.last_attempt_at)} · 最近成功 ${cacheTime(d.last_success_at)}</p></details></div>`;
}
async function loadDataStatus() {
  const root = document.getElementById('data-status');
  if (!root) return;
  try {
    const data = await getJSON('/api/data-status');
    const warnings = data.sources.filter(s => s.status !== 'ok').length;
    const stockWarnings = data.stocks.filter(s => !['current','pending'].includes(s.daily.status) || s.snapshot.status !== 'ok' || s.snapshot.freshness !== 'cached').length;
    root.querySelector('summary').textContent = `数据状态 · ${warnings || stockWarnings ? warnings + ' 类来源 / ' + stockWarnings + ' 只股票需留意' : '查看各来源与个股缓存'}`;
    root.querySelector('.cache-body').innerHTML = `<p class="muted">采集结果是来源汇总；个股缓存可能更早。全部内容来自本地数据库。</p>
      <div class="cache-source-grid">${data.sources.map(s => `<article class="cache-state${cacheNotice(s.status)}"><b>${esc(s.label)} · ${esc(cacheLabel(s.status))}</b>
      <p>来源内最新数据 ${esc(s.data_as_of || '业务时间未知')}</p><p>最近尝试 ${cacheTime(s.last_attempt_at)}</p><p>最近成功 ${cacheTime(s.last_success_at)}</p></article>`).join('')}</div>
      <details><summary>个股数据日期与快照状态（${data.stocks.length}）</summary><div class="cache-source-grid">${data.stocks.map(s => `<article class="cache-state"><b>${esc(s.code)}</b>
      <p>日线 ${esc(s.daily.data_as_of || '未知')} · ${esc(cacheLabel(s.daily.status))}</p>
      <p>快照 ${esc(cacheLabel(s.snapshot.status))} · ${esc(cacheLabel(s.snapshot.freshness))}</p></article>`).join('')}</div></details>
      ${data.truncated ? '<p>仅展示前 400 个标的，请在个股详情查看其余标的。</p>' : ''}`;
  } catch (_) {
    root.querySelector('summary').textContent = '数据状态 · 暂时不可用';
    root.querySelector('.cache-body').textContent = '未能读取数据状态，请稍后重试。';
  }
}
function renderCacheSnapshot(data) {
  const s=data.snapshot, v=s.values;
  const suspension = v.suspension === 1 ? '停牌' : v.suspension === 0 ? '未停牌' : '未知';
  const secStatus = String(v.sec_status || '').split('.').pop() === 'NORMAL' ? '正常' : '未知';
  const sourceZone = {'Asia/Hong_Kong':'香港时间','America/New_York':'美东时间'}[s.source_timezone] || '时区未知';
  const rows=[['开盘',v.open_price],['日内最高',v.high_price],['日内最低',v.low_price],['昨收',v.prev_close_price],['量比',v.volume_ratio],['每手股数',v.lot_size]];
  return cacheDaily(data.daily) + `<section class="cache-card">
    <div class="cache-heading"><h3>Futu 缓存快照</h3><span>${esc(s.currency)}</span></div>
    <p class="cache-state${cacheNotice(s.status)}">${esc(cacheLabel(s.status))} · ${esc(cacheLabel(s.freshness))}${s.reason ? ' · '+esc(s.reason) : ''}</p>
    ${s.migration_required ? '<p class="cache-warning">快照展示需要维护者更新数据库结构并采集数据。</p>' : ''}
    <div class="cache-price">${s.has_cache ? fmtNum(v.last_price) : '—'} <small class="${s.change === null ? '' : plClass(s.change)}">${s.change === null ? '涨跌未知' : fmtSigned(s.change)+' ('+fmtSigned(s.change_pct)+'%)'}</small></div>
    <p class="muted">来源时间 ${esc(s.data_as_of || '未知')} <span class="cache-zone">${sourceZone}</span><br>采集时间 ${cacheTime(s.collected_at)}</p>
    <dl class="cache-values">${rows.map(([label,value])=>`<div><dt>${esc(label)}</dt><dd>${label === '每手股数' ? fmtInt(value) : fmtNum(value)}</dd></div>`).join('')}
    <div><dt>停牌状态</dt><dd>${suspension}</dd></div><div><dt>证券状态</dt><dd>${secStatus}</dd></div></dl>
    <details><summary>来源与采集说明</summary><p>最近尝试 ${cacheTime(s.last_attempt_at)}<br>最近完整成功 ${cacheTime(s.last_success_at)}</p>
    <p>供应商证券状态 ${esc(v.sec_status || '未知')}。当前向上报价档位间隔 ${fmtNum(v.price_spread)}；仅为本次观察值，不用于历史交易规则。此卡为缓存，不是实时行情。</p></details>
    </section><div class="cache-state${cacheNotice(data.profile.status)}"><b>yfinance 公司资料 · ${esc(cacheLabel(data.profile.status))}</b><p>该股采集 ${cacheTime(data.profile.collected_at)} · 业务截至时间未知</p><p>最近来源尝试 ${cacheTime(data.profile.last_attempt_at)} · 最近来源成功 ${cacheTime(data.profile.last_success_at)}（${data.profile.scope === 'stock' ? '该股' : '来源汇总'}）</p></div>`;
}
async function loadCacheSnapshot(code) {
  const host = document.getElementById('detail-cache');
  try {
    const data=await getJSON(`/api/stock/${encodeURIComponent(code)}/snapshot`);
    if (host && host.isConnected) host.innerHTML=renderCacheSnapshot(data);
  } catch (_) { if(host && host.isConnected) host.textContent='缓存状态暂时不可用，其他已入库数据仍可查看。'; }
}
loadDataStatus();

File: mystock/web/data_status.py
"""Read-only cache provenance. No collectors, schema initialization or writes."""
from datetime import datetime
from zoneinfo import ZoneInfo
import re
from ..ml import sessions

SOURCES = {
    'futu_position': '持仓快照', 'futu_funds': '账户资金',
    'futu_order': '订单', 'futu_deal': '成交', 'yfinance': '日线行情',
    'fx_usdcny': '美元汇率', 'yf_profile': '公司资料',
    'futu_snapshot': 'Futu 缓存快照', 'futu_capflow': '资金流向',
}
REASONS = {
    'request_failed': '请求失败，保留上次缓存', 'not_returned': '本次未返回该标的',
    'missing_core_fields': '来源未提供必要字段', 'missing_fields': '字段不完整，保留上次缓存',
    'invalid_price': '价格无效，保留上次缓存', 'unknown_market': '市场无法识别',
    'unknown_source_time': '来源时间无法确认',
}


def table(conn, name):
    return bool(conn.execute("SELECT 1 FROM sqlite_master WHERE type='table' AND name=?", (name,)).fetchone())


def valid_code(code):
    return bool(re.fullmatch(r'(HK|US)\.[A-Za-z0-9._-]{1,24}', code))


def stamp(value):
    """Old naive timestamps retain their text, never an invented time zone."""
    try:
        dt = datetime.fromisoformat(value.replace('Z', '+00:00'))
        return dt.isoformat() if dt.tzinfo is not None else None
    except (AttributeError, TypeError, ValueError):
        return None


def aggregate(conn, source):
    rows = []
    if table(conn, 'sync_log'):
        for condition in ('', " AND status='ok'"):
            row = conn.execute('SELECT status,run_at FROM sync_log WHERE source=?' + condition + ' ORDER BY id DESC LIMIT 1', (source,)).fetchone()
            rows.append(dict(row) if row else {})
    else:
        rows = [{}, {}]
    latest, success = rows
    time_columns = {
        'futu_position': ('positions', 'snapshot_date'), 'futu_funds': ('account_funds', 'snapshot_date'),
        'futu_order': ('orders', 'updated_time'), 'futu_deal': ('deals', 'create_time'),
        'yfinance': ('daily_quotes', 'date'), 'fx_usdcny': ('fx_rates', 'date'),
        'futu_capflow': ('capital_flow', 'date'), 'futu_snapshot': ('stock_profiles', 'snapshot_time_utc'),
    }
    data_as_of = None
    if source in time_columns:
        name, column = time_columns[source]
        if column in {r[1] for r in conn.execute(f'PRAGMA table_info({name})')}:
            data_as_of = conn.execute(f'SELECT MAX({column}) FROM {name}').fetchone()[0]
    return dict(source=source, label=SOURCES[source], scope='source_summary',
                status=latest.get('status', 'unknown'), data_as_of=data_as_of,
                last_attempt_at=latest.get('run_at'), last_success_at=success.get('run_at'),
                attempt_timezone_known=stamp(latest.get('run_at')) is not None)


def daily_state(code, row, now):
    result = dict(status='unknown', expected_session=None, data_as_of=row.get('date') if row else None)
    try:
        if not valid_code(code):
            return result
        current = sessions.utc(now)
        zone = ZoneInfo('Asia/Hong_Kong' if code.startswith('HK.') else 'America/New_York')
        day = current.astimezone(zone).date().isoformat()
        dates = sessions.window(code, day, 2)
        closed = [d for d in dates if sessions.session(code, d)['final_at'] <= current]
        if not closed:
            return result
        expected = closed[-1]
        result['expected_session'] = expected
        if not row:
            result['status'] = 'empty'
        elif row['date'] < expected:
            result['status'] = 'stale'
        elif row['date'] > day:
            result['status'] = 'unknown'
        elif row['date'] > expected:
            result['status'] = 'pending'
        elif not stamp(row.get('synced_at')):
            result['status'] = 'unknown'
        elif sessions.utc(row['synced_at']) > current:
            result['status'] = 'unknown'
        elif sessions.daily_final(code, row, current):
            result['status'] = 'current'
        else:
            result['status'] = 'awaiting_final_data'
    except (ValueError, KeyError):
        pass
    return result


def stock_attempt(conn, source, code):
    if not table(conn, 'collection_status'):
        return None
    row = conn.execute('SELECT * FROM collection_status WHERE source=? AND code=?', (source, code)).fetchone()
    if not row:
        return None
    row = dict(row)
    row['scope'] = 'stock'
    row['reason'] = REASONS.get(row.get('reason'))
    return row


def stock_status(conn, code, now, summaries=None):
    summaries = summaries or {s: aggregate(conn, s) for s in ('yfinance', 'yf_profile')}
    quote = conn.execute('SELECT * FROM daily_quotes WHERE futu_code=? ORDER BY date DESC LIMIT 1', (code,)).fetchone() if table(conn, 'daily_quotes') else None
    profile = conn.execute('SELECT * FROM stock_profiles WHERE futu_code=?', (code,)).fetchone() if table(conn, 'stock_profiles') else None
    profile = dict(profile) if profile else {}
    q = dict(quote) if quote else None
    daily = {**summaries['yfinance'], **daily_state(code, q, now),
             'collected_at': q.get('synced_at') if q else None}
    quote_attempt = stock_attempt(conn, 'yfinance', code) or summaries['yfinance']
    daily.update({k: quote_attempt[k] for k in ('scope','last_attempt_at','last_success_at')})
    daily['attempt_status'] = quote_attempt['status']
    if daily['attempt_status'] != 'ok' and daily['status'] == 'current':
        daily['status'] = 'cached'
    details = {**summaries['yf_profile'], 'collected_at': profile.get('synced_at'),
               'data_as_of': None, 'has_cache': bool(profile.get('synced_at'))}
    profile_attempt = stock_attempt(conn, 'yf_profile', code)
    if profile_attempt:
        details.update(profile_attempt)
    if not details['has_cache']:
        details['status'] = 'empty'
    attempt = None
    migrated = 'snapshot_time_utc' in {r[1] for r in conn.execute('PRAGMA table_info(stock_profiles)')}
    if table(conn, 'collection_status'):
        attempt = conn.execute("SELECT * FROM collection_status WHERE source='futu_snapshot' AND code=?", (code,)).fetchone()
    attempt = dict(attempt) if attempt else {}
    snapshot = dict(source='futu_snapshot', scope='stock', status=attempt.get('status', 'unknown'),
                    last_attempt_at=attempt.get('last_attempt_at'), last_success_at=attempt.get('last_success_at'),
                    collected_at=profile.get('snap_synced_at'), data_as_of=profile.get('snapshot_time_raw'),
                    source_timezone=profile.get('snapshot_timezone'), source_time_utc=profile.get('snapshot_time_utc'),
                    reason=REASONS.get(attempt.get('reason')), migration_required=not migrated)
    values = {k: profile.get(k) for k in ('last_price','prev_close_price','open_price','high_price','low_price','volume_ratio','suspension','sec_status','lot_size','price_spread')}
    last, prev = values['last_price'], values['prev_close_price']
    snapshot['values'] = values
    snapshot['currency'] = 'HKD' if code.startswith('HK.') else 'USD'
    snapshot['change'] = last - prev if last is not None and prev is not None and prev > 0 else None
    snapshot['change_pct'] = snapshot['change'] / prev * 100 if snapshot['change'] is not None else None
    snapshot['has_cache'] = last is not None and bool(profile.get('snapshot_time_utc'))
    snapshot['freshness'] = 'unknown'
    try:
        current = sessions.utc(now)
        source = sessions.utc(profile['snapshot_time_utc'])
        observed = sessions.utc(profile['snap_synced_at'])
        if source > observed or observed > current:
            raise ValueError('future clock')
        zone = ZoneInfo(snapshot['source_timezone'])
        day = current.astimezone(zone).date().isoformat()
        dates = sessions.window(code, day, 2)
        active = next((sessions.session(code, d) for d in dates if sessions.session(code, d)['open'] <= current < sessions.session(code, d)['final_at']), None)
        closed = [sessions.session(code, d) for d in dates if sessions.session(code, d)['final_at'] <= current]
        cutoff = active['open'] if active else closed[-1]['open']
        # Cached quote time need not equal the closing auction time (suspensions,
        # illiquid securities). Show its exact time; no claim of real-time prices.
        snapshot['freshness'] = 'cached' if source >= cutoff else 'stale'
    except (ValueError, TypeError, KeyError, IndexError):
        pass
    return dict(code=code, daily=daily, profile=details, snapshot=snapshot)


def overview(conn, now):
    codes = set()
    for name in ('positions', 'orders', 'deals'):
        if table(conn, name):
            codes.update(r[0] for r in conn.execute(f'SELECT DISTINCT code FROM {name}') if valid_code(r[0] or ''))
    # Bounded response; aggregate collection status does not imply per-stock health.
    codes = sorted(codes)
    summaries = {s: aggregate(conn, s) for s in SOURCES}
    return dict(schema_version=1, sources=list(summaries.values()),
                stocks=[stock_status(conn, code, now, summaries) for code in codes[:400]], truncated=len(codes)>400)
